js/client-mainmenu.js

```diff
@@ -1487,5 +1500,151 @@
+	var on_team_select=new Event("onselectteam_event");
+
+	function attatchEventToElement(element_name,event_name,even_lisener){
+		let element=document.querySelector(element_name)
+		element.addEventListener(event_name,function(e){even_lisener()});
+	}
+	function trigerEvent(element_name,event_obj){
+		let element=document.querySelector(element_name)
+		element.dispatchEvent(event_obj)
+	}
+	//creat observer for log in fourm to check when user log in 
+	function createloginFourmObserver(){
+		const config = { attributes: true, childList: true, subtree: true };
+		const observer = new MutationObserver(function(){
+		if(window.app.user.attributes.named && window.localStorage.page=="challenge"){
+			window.localStorage.page="testclient";
+			if($('.challenge-container').length==0){
+			//user log in event trigerd when user log in and make user pick a team 
+			var on_user_login_event=new Event("login_event");
+			attatchEventToElement('body','login_event',onUserLogIn);
+			trigerEvent('body',on_user_login_event);
+			}
+		}
+		});
+		observer.observe(document.querySelector('body'), config);
+	}
+	
+	 function onUserLogIn(){
+		var buf = '<div class="challenge"><form class="battleform"><p>Challenge ' + BattleLog.escapeHTML("xdxdblack") + '?</p>';
+			buf += '<p><label class="label">Format:</label>' + new MainMenuRoom().renderFormats(window.localStorage.format) + '</p>';
+			buf += '<p id="select-team"><label class="label">Team:</label>' + new MainMenuRoom().renderTeams(window.localStorage.format) + '</p>';
+			buf +='<button type="button" id=accept-challenge>accept</button>';
+			buf +='<button type="button" id=close-challenge>close</button></form></div>'
+			$('.pmbox').append(`<div class=challenge-container>${buf}</div>`);
+			$('.select.formatselect').prop("disabled",true);
+			$('.select.teamselect').prop("disabled",false);
+			$('#accept-challenge').click(function(){
+				onTeamSelect(window.localStorage.challenger).then(function({massage,userid}){
+					 $('#accept-challenge').html("accept");
+					if(massage.includes("Your team is valid") && userid!=app.user.attributes.name && userid){
+						sendchallenge({userid:userid,format:$('.select.formatselect').val()});
+						$('.challenge-container').empty();
+					 }
+					 else if(userid==app.user.attributes.name){
+						alert("userid and challeger id are equall you cant challenge your self");
+					 }
+					else{
+						alert(massage);
+					}
+				})
+			})
+			$('#close-challenge').click(function(){
+				$('.challenge-container').empty();
+			})
+	}
+	  function onTeamSelect(challenger_name){
+		return new Promise(function(resolve){
+		   $('#accept-challenge').html("processing request...");
+			if(window.localStorage.team!="random"){
+			app.sendTeam(window.Storage.teams[$('#select-team .select.teamselect').val()]);
+		   app.send('/vtm ' + $('.select.formatselect').val());
+			setTimeout(function(){
+				resolve({userid:challenger_name,massage:window.localStorage.errormassege});
+			},1000)
```

Comment:
1000ms is not necessarily accurate. It would probably be worth either having the server send a protocol message that the validation is done, or you could hardcode a check for popups.